### DISCUSSION TITLE
Unable to write HDDL files from Hierarchical problems

### DISCUSSION THREAD

**Discussion by Elliot1008** [aiplan4eu/unified-planning#603] [Q&A] [answered]
It is my understanding that we use the same PDDL writer for hierarchical problems as well?
However, when I try to write my hierarchical problem, it attempts to do it as PDDL and not HDDL, and then it throws an error:

`unified_planning.exceptions.UPTypeError: PDDL supports only boolean and numerical fluents`

How do I convert my hierarchical problem into domain and problem files?

**Comment by arbimo** [comment]
Despite the different name, HDDL is de facto an extension of PDDL with HTN components. So any limitation of PDDL also applies to HDDL.

So if you want to convert it to HDDL/PDDL, the problem needs to be specified in a compatible form. Notably, this means that it should not contain fluents that return a `UserType` (which seems to be the source of your problem above.

**Comment by Elliot1008** [comment]
Okay, as instructed, I converted the hierarchical planning problem to only have boolean fluents. While the writer can now convert it without throwing errors, it captured only the PDDL elements (actions, predicates), it seems to have lost all the methods I defined!  Is this expected behavior?

Ideally, I'd like the hddl file to retain the methods structure along with the PDDL elements.

**Comment by arbimo** [comment, marked as answer]
HDDL support was only added recently (#577) and may not be in the released version yet. Have you tried using the latest version on master?

**Comment by Elliot1008** [comment]
Yes, this seemed to do the trick. Thanks
